src/util/submit-addon.js

```
@@ -0,0 +1,326 @@
+        if (file && file.status === 'public') {
+          return file.url;
+        }
+        return null;
+      },
+      `${this.apiUrl}addon/${addonId}/versions/${versionId}/`,
+      this.approvalCheckInterval,
+      this.approvalCheckTimeout,
+    );
+  }
+
+  async fetchJson(
+    url: string,
+    method: string = 'GET',
+    body?: typeof FormData | string,
+    errorMsg: string = 'Bad Request'
+  ): Promise<any> {
+    const response = await this.fetch(url, method, body);
+    if (response.status < 200 || response.status >= 500) {
+      throw new Error(
+        `${errorMsg}: ${response.statusText || response.status}.`);
+    }
+    const data = await response.json();
+
+    if (!response.ok) {
+      log.info('Server Response:', data);
+      throw new Error(
+        `${errorMsg}: ${response.statusText || response.status}.`);
+    }
+    return data;
+  }
+
+  async signJWT(): Promise<string> {
+    return new SignJWT({ iss: this.apiKey })
+      .setProtectedHeader({ alg: 'HS256' })
+      .setIssuedAt()
+      .setExpirationTime(`${this.apiJwtExpiresIn}seconds`)
+      .sign(Uint8Array.from(Buffer.from(this.apiSecret, 'utf8')));
+  }
+
+  async fetch(
+    url: string,
+    method: string = 'GET',
+    body?: typeof FormData | string,
+  ): Promise<typeof Response> {
+    const authToken = await this.signJWT();
+
+    log.info(`Fetching URL: ${url}`);
+    let headers = {
+      Authorization: `JWT ${authToken}`,
+      Accept: 'application/json',
+    };
+    if (typeof body === 'string') {
+      headers = {
+        ...headers,
+        'Content-Type': 'application/json',
+      };
+    }
+    return fetch(url, { method, body, headers });
+  }
+
+  async downloadSignedFile(
+    fileUrl: string,
+    addonId: string
+  ): Promise<SignResult> {
+    const filename = fileUrl.split('/').pop(); // get the name from fileUrl
+    const dest = `${this.downloadDir}/${filename}`;
+    const response = await this.fetch(fileUrl);
```

That might throw, are we OK with this here or should we use try/catch to return the same error as below ("Downloading xyz failed")?

---

mmm, do you under what circumstances it would throw rather than just return a response that is `! response.ok`?

---

maybe no internet connection (or hiccup)? I am not sure..

---

it now catches the error for the fetch and for the `saveToFile` call.